### PR TITLE
Show archived rooms and implement forgetting

### PR DESCRIPTION
**Pull request by Kegsay**
Part of fixing https://github.com/vector-im/vector-web/issues/348

Forget button:
 - Shown if the user has a membership of `leave` (and hides the "Leave Room" button)
 - When clicked hits the forget API which nukes it from the store.
 - You're then taken to another room.
 - *May require some new assets for a better icon. I reused the leave icon*.

Archived rooms:
 - All rooms with a membership of `leave` or `ban` will now be dumped to the "Historical" section which was already in place. Listen for when rooms are deleted so we can update this list when people forget rooms.
 - Make the `RoomSubList` show a spinner if we are currently loading left rooms. This is done via the `showSpinner` prop.
 - Make the `RoomSubList` always display the archived section (because we don't know beforehand if there are archived rooms!). This is done via the `alwaysShowHeader` prop.
 - Call `syncLeftRooms()` when the archived header is clicked. This required the prop `startAsHidden` to be added to force the archived section to be hidden on start (so clicking it *once* will trigger the request).

Performance:
When the JS SDK encounters a new room it will emit a flurry of events for things
like state and room members. Refreshing the room list on each event is bad for
performance. This is okay initially because the room list is only shown after
the first sync, but when getting archived rooms it locks up for 15-30s as it
thrashes. Add a 1s cap to refreshRoomList() which means that it will refresh
*AT MOST* once every second. If it has been >1s since the last refresh it will
immediately refresh. If it has been <1s it will wait the difference. This improves
performance and results in 3-5s lock. More optimisations can be done in future.

Manually tested and it all works.

See also:
 - https://github.com/matrix-org/matrix-js-sdk/pull/56
 - https://github.com/vector-im/vector-web/pull/541

### CODE REVIEW

File: src/components/structures/RoomView.js [hunk 1]
```diff
@@ -865,7 +865,14 @@ module.exports = React.createClass({
...
             room_id: this.props.roomId,
         });
-        this.props.onFinished();        
+    },
+
+    onForgetClick: function() {
+        MatrixClientPeg.get().forget(this.props.roomId).done(function() {
+            dis.dispatch({ action: 'view_next_room' });
+        }, function(err) {
+            console.error("Failed to forget room: %s", err);
```
Comment (ara4n) [inline src/components/structures/RoomView.js]: why would we swallow this error msg? can we ErrorDialog it please?

File: src/components/views/rooms/RoomList.js [hunk 2]
```diff
@@ -45,6 +46,7 @@ module.exports = React.createClass({
     componentWillMount: function() {
         var cli = MatrixClientPeg.get();
         cli.on("Room", this.onRoom);
+        cli.on("deleteRoom", this.onDeleteRoom);
```
Comment (ara4n) [inline src/components/views/rooms/RoomList.js]: why isn't this Room.delete or something?

Comment (Kegsay) [inline src/components/views/rooms/RoomList.js]: `Room.whatever` is by convention used to tell people that the property `whatever` on the provided `Room` object has changed. This isn't the case here.

**Comment by ara4n** [comment]
LGTM, other than the fact that we've lost the "bottommost" CSS somehow for the Archived RoomSubList.  This is meant to provide a whacking great margin on the bottom of the sublist to avoid the browser adjusting the scrollTop of the RoomList when users toggle sublists on and off (as otherwise if you hide the bottommost ones, it will scroll you upwards to make the bottommost list touch the bottom of the scroll area).  It was also serving as a larger droptarget ,although this is less relevant given you can't currently dnd to archived (i assume).  Would be nice to get it back - either by putting the margin on the overall roomlist or back on the bottommost roomsublist.

File: src/components/structures/RoomView.js (hunk 1 above)
Comment (Kegsay) [inline src/components/structures/RoomView.js]: Done

**Comment by Kegsay** [comment]
PTAL

**Comment by ara4n** [comment]
lgtm, thanks.

**Pull request merged**